org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/builderState/AbstractBuilderState.java

```diff
@@ -52,22 +58,48 @@
 	
 	@Inject
 	private IStorage2UriMapper storage2UriMapper;
+	
+	@Inject
+	private IWorkspace workspace;
 
 	private volatile boolean isLoaded = false;
 
-	public synchronized void load() {
+	public void load() {
 		if (!isLoaded) {
-			resourceDescriptionData = new ResourceDescriptionsData(persister.load());
-			if(storage2UriMapper instanceof IStorage2UriMapperExtension)
-				((IStorage2UriMapperExtension) storage2UriMapper).initializeCache();
-			isLoaded = true;
+			IWorkspaceRunnable runnable = new IWorkspaceRunnable() {
+				@Override
+				public void run(IProgressMonitor monitor) throws CoreException {
+					if (!isLoaded) {
+						resourceDescriptionData = new ResourceDescriptionsData(persister.load());
+						if(storage2UriMapper instanceof IStorage2UriMapperExtension)
+							((IStorage2UriMapperExtension) storage2UriMapper).initializeCache();
+						isLoaded = true;
+					}
+				}
+			};
+			try {
+				if (isWorkspaceLocked()) {
+					runnable.run(null);
+				} else {
+					workspace.run(runnable, null, IWorkspace.AVOID_UPDATE, null);
+				}
+			} catch(CoreException e) {
+				log.error(e.getMessage(), e);
+			}
+		}
+	}
+	
+	@SuppressWarnings("restriction")
+	private boolean isWorkspaceLocked() {
+		try {
+			return ((org.eclipse.core.internal.resources.Workspace) workspace).getWorkManager().isLockAlreadyAcquired();
+		} catch (CoreException e) {
+			return false;
```

Comment:
_workManager may become null on shutdown. = > CoreExceptionso i am thinking if returning false is correct here